pandas/core/indexes/period.py

```diff
@@ -441,7 +442,9 @@ def _formatter_func(self):
 
     @cache_readonly
     def _engine(self):
-        return self._engine_type(lambda: self, len(self))
+        # To avoid a reference cycle, pass a weakref of self to _engine_type.
+        period = weakref.ref(self)
```

Comment:
should we be using weakref in the other invocations for consistency?

Reply:
weakrefs cause some tests to fail.

I'm not sure currently if those failures are correct, could be worth an investigation. @crepererum , did you investigate if weakrefs not working is the correct behaviour?

Reply:
No, I didn't test this.

IMHO, the weakref isn't required in the other cases since we're passing a member of `self` and not `self` itself. Using a weakref seemed to make things only unnecessary complicated, since then you have just a chain of indirections (`lambda->scope->weakref->self->member` instead of `lambda->scope->member`).